astropy/table/index.py

```diff
@@ -852,14 +847,41 @@ def _get_rows(self, item):
                 item = [item]
             # item should be a list or ndarray of values
             rows = []
-            for key in item:
-                p = index.find((key,))
+            for index_key in item:
+                p = index.find((index_key,))
                 if len(p) == 0:
-                    raise KeyError(f"No matches found for key {key}")
+                    raise KeyError(f"No matches found for key {index_key}")
                 else:
                     rows.extend(p)
         return rows
 
+    def _get_row_idxs_as_list_or_int(self, item):
+        """Internal function to retrieve row indices for ``item`` as a list or int.
+
+        See ``__getitem__`` for details on the input item.
+        """
+        # This handles ``tbl.loc[<item>]`` and ``tbl.loc[<key>, <item>]`` (and the same
+        # for ``tbl.loc_indices``).
+        if isinstance(item, tuple):
+            key, item = item
+        else:
+            key = self.table.primary_key
+
+        item_is_sequence = isinstance(item, (list, np.ndarray))
+
+        # Short-circuit for case like tbl.loc[[]], returns tbl[[]]
+        if item_is_sequence and len(item) == 0:
+            return []
+
+        rows = self._get_row_idxs_as_list(key, item)
+        # If ``item`` is a sequence of keys or a slice then always returns a list of
+        # rows, where zero rows is OK. Otherwise check output and possibly return a
+        # scalar.
+        if not (item_is_sequence or isinstance(item, slice)) and len(rows) == 1:
+            rows = rows[0]
```

Comment:
This code here constitutes the real functional changes. The other changes are basically just refactoring.